src/main/scala/fi/oph/koski/kela/KelaAmmatillinenSchema.scala

```diff
@@ -0,0 +1,178 @@
+import fi.oph.scalaschema.annotation.{Description, Title}
+
+import java.time.{LocalDate, LocalDateTime}
+
+@Title("Ammatillisen koulutuksen opiskeluoikeus")
+@Description("Ammatillisen koulutuksen opiskeluoikeus")
+case class KelaAmmatillinenOpiskeluoikeus(
+  oid: Option[String],
+  versionumero: Option[Int],
+  aikaleima: Option[LocalDateTime],
+  oppilaitos: Option[Oppilaitos],
+  koulutustoimija: Option[Koulutustoimija],
+  sisältyyOpiskeluoikeuteen: Option[Sisältäväopiskeluoikeus],
+  arvioituPäättymispäivä: Option[LocalDate],
+  ostettu: Option[Boolean],
+  tila: KelaAmmatillisenOpiskeluoikeudenTila,
+  suoritukset: List[KelaAmmatillinenPäätasonSuoritus],
+  lisätiedot: Option[KelaAmmatillisenOpiskeluoikeudenLisätiedot],
+  @KoodistoKoodiarvo("ammatillinenkoulutus")
+  tyyppi: schema.Koodistokoodiviite,
+  organisaatioHistoria: Option[List[OrganisaatioHistoria]]
+) extends KelaOpiskeluoikeus {
+  override def alkamispäivä: Option[LocalDate] = super.alkamispäivä
+  override def päättymispäivä: Option[LocalDate] = super.päättymispäivä
+  def withEmptyArvosana: KelaAmmatillinenOpiskeluoikeus = copy(
+    suoritukset = suoritukset.map(_.withEmptyArvosana)
+  )
+}
+
+case class KelaAmmatillisenOpiskeluoikeudenTila(
+  opiskeluoikeusjaksot: List[KelaAmmatillisenOpiskeluoikeusjakso]
+) extends OpiskeluoikeudenTila
+
+case class KelaAmmatillisenOpiskeluoikeusjakso(
+  alku: LocalDate,
+  tila: KelaKoodistokoodiviite,
+  opintojenRahoitus: Option[KelaKoodistokoodiviite]
+) extends Opiskeluoikeusjakso
+
+case class KelaAmmatillisenOpiskeluoikeudenLisätiedot(
+  majoitus: Option[List[schema.Aikajakso]],
+  sisäoppilaitosmainenMajoitus: Option[List[schema.Aikajakso]],
+  @SensitiveData(Set(Rooli.LUOTTAMUKSELLINEN_KELA_LAAJA))
+  vaativanErityisenTuenYhteydessäJärjestettäväMajoitus: Option[List[schema.Aikajakso]],
+  @SensitiveData(Set(Rooli.LUOTTAMUKSELLINEN_KELA_LAAJA))
+  erityinenTuki: Option[List[schema.Aikajakso]],
+  @SensitiveData(Set(Rooli.LUOTTAMUKSELLINEN_KELA_LAAJA))
+  vaativanErityisenTuenErityinenTehtävä: Option[List[schema.Aikajakso]],
+  ulkomaanjaksot: Option[List[Ulkomaanjakso]],
+  @SensitiveData(Set(Rooli.LUOTTAMUKSELLINEN_KELA_LAAJA))
+  hojks: Option[Hojks],
+  osaAikaisuusjaksot: Option[List[schema.OsaAikaisuusJakso]],
+  opiskeluvalmiuksiaTukevatOpinnot: Option[List[schema.OpiskeluvalmiuksiaTukevienOpintojenJakso]],
+  vankilaopetuksessa: Option[List[schema.Aikajakso]],
+  maksuttomuus: Option[List[schema.Maksuttomuus]],
+  oikeuttaMaksuttomuuteenPidennetty: Option[List[schema.OikeuttaMaksuttomuuteenPidennetty]]
+) extends OpiskeluoikeudenLisätiedot
+
+@Title("Ammatillisen koulutuksen suoritus")
+case class KelaAmmatillinenPäätasonSuoritus(
+  koulutusmoduuli: KelaAmmatillisenSuorituksenKoulutusmoduuli,
+  suoritustapa: Option[KelaKoodistokoodiviite],
+  toimipiste: Option[Toimipiste],
+  vahvistus: Option[Vahvistus],
+  osasuoritukset: Option[List[KelaAmmatillinenOsasuoritus]],
+  tyyppi: schema.Koodistokoodiviite,
```

Comment:
Miksi osassa tyyppi-kenttiä on `@KoodistoKoodiarvo`-annotaatio, ja osassa ei? Mietin, että jääköhän jotain tapauksia, jotka ei vaan sitten tuotannossa deserialisoidu, vaikka testeissä olevat opiskeluoikeudet meneekin läpi ilman näitä annotaatioita?

Reply:
Mielestäni `@KoodistoKoodiarvo` annotaatiota tarvitaan serialisaatiolle vihjeeksi vain "anyOf"-tapauksissa eli kun jonkin kentän tyyppinä on `trait` josta periytyy muita luokkia. Diskriminator ja Koodistokoodiarvo -annotaatiot yhdessä antavat vihjeen serialisointiin, mitä konkreettista luokkaa pitäisi käyttää.

Esimerkiksi tässä tapauksessa kuitenkin `KelaAmmatillinenOpiskeluoikeus` käyttää suoraan konkreettista tyyppiä `suoritukset: List[KelaAmmatillinenPäätasonSuoritus],` jolloin serialisaation pitäisi aina osata serialisoida suoritukset käyttäen luokkaa `KelaAmmatillinenPäätasonSuoritus`.